pyccel/ast/mathext.py

```diff
@@ -13,6 +13,7 @@
     'math_functions',
     # ---
     'MathFunctionBase',
+    'PycMathFunctionBase',
```

Comment:
Can you explain the difference between these two classes?

Reply:
MathFunctionBase is the abstract for the math functions that are already implemented in C `math.h` library.
the other one is for `pyc_math `library functions.

Reply:
Hmmm, I'm not sure this separation is necessary. Especially before the codegen stage as different languages have different math functions already implemented. If I have understood correctly the only difference between `MathFunctionBase` and `PycMathFunctionBase` is that you need to import `pyc_math.h` for `PycMathFunctionBase` and `math.h` for `MathFunctionBase`.
So `_print_MathFunctionBase` would be:
```
        self._additional_imports.add('pyc_math')
        type_name = type(expr).__name__
        try:
            func_name = pyc_math_dic[type_name]
        except KeyError:
            errors.report(PYCCEL_RESTRICTION_TODO, severity='fatal')
        args = []
        for arg in expr.args:
            if arg.dtype != expr.dtype:
                cast_func = python_builtin_datatypes[str_dtype(expr.dtype)]
                args.append(self._print(cast_func(arg)))
            else:
                args.append(self._print(arg))
        code_args = ', '.join(args)
        return '{0}({1})'.format(func_name, code_args)
```
and `_print_PycMathFunctionBase` would be:
```
        self._additional_imports.add('pyc_math')
        type_name = type(expr).__name__
        try:
            func_name = pyc_math_dic[type_name]
        except KeyError:
            errors.report(PYCCEL_RESTRICTION_TODO, severity='fatal')
        args = []
        for arg in expr.args:
            if arg.dtype != expr.dtype:
                cast_func = python_builtin_datatypes[str_dtype(expr.dtype)]
                args.append(self._print(cast_func(arg)))
            else:
                args.append(self._print(arg))
        code_args = ', '.join(args)
        return '{0}({1})'.format(func_name, code_args)
```
can't these be merged to:
```
        type_name = type(expr).__name__
        try:
            func_name, import_file = pyc_math_dic[type_name]
        except KeyError:
            errors.report(PYCCEL_RESTRICTION_TODO, severity='fatal')
        self._additional_imports.add(import_file)
        args = []
        for arg in expr.args:
            if arg.dtype != expr.dtype:
                cast_func = python_builtin_datatypes[str_dtype(expr.dtype)]
                args.append(self._print(cast_func(arg)))
            else:
                args.append(self._print(arg))
        code_args = ', '.join(args)
        return '{0}({1})'.format(func_name, code_args)
```
or similar?

Reply:
this seems good. there is some differentiation, but I will try to make a general structure that can handle all these cases.

Reply:
That would be great. If it's just one or two classes that need handling differently (e.g MathFloor) then they can still have their own specialised method. But I'm pretty sure most of these can be handled really generally using `MathFunctionInt`, `MathFunctionFloat`, etc

Reply:
This seems beyond the scope of this PR. If these specific classes can't be handled by the existing `_print_MathFunctionBase` then I would suggest giving them specialised implementations (i.e. `_print_MathLcm`) and leaving the math printing improvement for another PR. It would be good if you could open an issue describing the changes that must be made. I would rather use specialised implementations than create a `PycFunctionBase` class as this makes no sense to me outside of the codegen stage